presto-hive/src/test/java/io/prestosql/plugin/hive/metastore/TestSemiTransactionalHiveMetastore.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.hive.metastore;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.plugin.hive.authentication.HiveIdentity;
+import org.testng.annotations.Test;
+
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.IntStream;
+
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.prestosql.plugin.hive.HiveTestUtils.HDFS_ENVIRONMENT;
+import static io.prestosql.testing.TestingConnectorSession.SESSION;
+import static java.lang.Math.abs;
+import static java.util.concurrent.Executors.newFixedThreadPool;
+import static org.testng.Assert.assertTrue;
+
+public class TestSemiTransactionalHiveMetastore
+{
+    private static final int SLEEP_TIME = 1_000;
+    private static final int PARTITIONS_TO_DROP = 3;
+    private static final IntStream DROP_THREADS_CONFIG = IntStream.of(1, PARTITIONS_TO_DROP);
+
+    @Test
+    public void testParallelPartitionDrops()
+    {
+        DROP_THREADS_CONFIG.forEach(dropThreads -> {
+            SemiTransactionalHiveMetastore semiTransactionalHiveMetastore = getSemiTransactionalHiveMetastoreWithDropExecutor(newFixedThreadPool(dropThreads));
+            IntStream.range(0, PARTITIONS_TO_DROP).forEach(i -> semiTransactionalHiveMetastore.dropPartition(SESSION,
+                    "test",
+                    "test",
+                    ImmutableList.of(String.valueOf(i))));
+
+            long startTime = System.nanoTime();
+            semiTransactionalHiveMetastore.commit();
+            long timeElapsed = TimeUnit.MILLISECONDS.convert(System.nanoTime() - startTime, TimeUnit.NANOSECONDS);
+            int expectedTimeElapsed = (PARTITIONS_TO_DROP / dropThreads) * SLEEP_TIME;
+
+            assertTrue(abs(timeElapsed - expectedTimeElapsed) < SLEEP_TIME / 2,
```

Comment:
this is a very fragile condition. Instead I would count `dropPartition` from different threads, but only allow previous drop to finish once the next thread initializes drop